apis/oauth/src/main/java/org/jclouds/oauth/v2/filters/ClientCredentialsSecretFlow.java

```diff
@@ -0,0 +1,100 @@
+ */
+package org.jclouds.oauth.v2.filters;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Supplier;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import org.jclouds.oauth.v2.AuthorizationApi;
+import org.jclouds.oauth.v2.domain.ClientSecret;
+import org.jclouds.oauth.v2.config.OAuthScopes;
+import org.jclouds.oauth.v2.domain.Token;
+import org.jclouds.domain.Credentials;
+import org.jclouds.http.HttpException;
+import org.jclouds.http.HttpRequest;
+import org.jclouds.location.Provider;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.jclouds.Constants.PROPERTY_SESSION_INTERVAL;
+import static org.jclouds.oauth.v2.config.OAuthProperties.RESOURCE;
+
+/**
+ * Authorizes new Bearer Tokens at runtime by sending up for the http request.
+ *
+ * <h3>Cache</h3>
+ * This maintains a time-based Bearer Token cache. By default expires after 59 minutes
+ * (the maximum time a token is valid is 60 minutes).
+ * This cache and expiry period is system-wide and does not attend to per-instance expiry time
+ * (e.g. "expires_in" from Google Compute -- which is set to the standard 3600 seconds).
+ */
+public class ClientCredentialsSecretFlow implements OAuthFilter {
+    private static final Joiner ON_COMMA = Joiner.on(",");
+
+    private final Supplier<Credentials> credentialsSupplier;
+    private final OAuthScopes scopes;
+    private final long tokenDuration;
+    private final String resource;
+    private final LoadingCache<ClientSecret, Token> tokenCache;
+
+    @Inject ClientCredentialsSecretFlow(AuthorizeToken loader, @Named(PROPERTY_SESSION_INTERVAL) long tokenDuration,
+                                        @Named(RESOURCE) String resource,
+                                        @Provider Supplier<Credentials> credentialsSupplier, OAuthScopes scopes) {
+        this.credentialsSupplier = credentialsSupplier;
+        this.scopes = scopes;
+        this.resource = resource;
+        this.tokenDuration = tokenDuration;
+        // since the session interval is also the token expiration time requested to the server make the token expire a
+        // bit before the deadline to make sure there aren't session expiration exceptions
+        long cacheExpirationSeconds = tokenDuration > 30 ? tokenDuration - 30 : tokenDuration;
+        this.tokenCache = CacheBuilder.newBuilder().expireAfterWrite(cacheExpirationSeconds, SECONDS).build(loader);
+    }
+
+    static final class AuthorizeToken extends CacheLoader<ClientSecret, Token> {
+        private final AuthorizationApi api;
+
+        @Inject AuthorizeToken(AuthorizationApi api) {
+            this.api = api;
+        }
+
+        @Override public Token load(ClientSecret key) throws Exception {
+            return api.authorizeClientSecret(key.clientId(), key.clientSecret(), key.resource());
```

Comment:
According to [the spec](http://tools.ietf.org/html/rfc6749#section-4.4.2), the scope parameter is optional. It is already present in the `ClientSecret` object and we're configuring it in the Azure OAuth module. Is there any reason why it is not being sent in the request?